core/runtime/common/runtime_upgrade_tracker_impl.hpp

```diff
@@ -28,20 +28,39 @@ namespace kagome::runtime {
         std::shared_ptr<const blockchain::BlockHeaderRepository> header_repo);
 
     void subscribeToBlockchainEvents(
-        std::shared_ptr<primitives::events::StorageSubscriptionEngine>
-            storage_sub_engine,
+        std::shared_ptr<primitives::events::ChainSubscriptionEngine>
+            chain_sub_engine,
         std::shared_ptr<const blockchain::BlockTree> block_tree);
 
     outcome::result<storage::trie::RootHash> getLastCodeUpdateState(
         const primitives::BlockInfo &block) const;
 
    private:
+    bool isStateInChain(const primitives::BlockInfo &state,
+                        const primitives::BlockInfo &chain_end) const;
+
+    struct RuntimeUpgradeData {
+
+      template <typename BlockInfo, typename RootHash>
+      RuntimeUpgradeData(BlockInfo &&block, RootHash &&state)
```

Comment:
You have exact types. Why template is used here?

Reply:
To abuse universal references and avoid creating separate versions for l-value and r-value reference arguments. Actually, my head hurt when I wrote this, so if you have a better idea I would be very welcome.